discovery_engine_core/core/src/storage/sqlite.rs

```diff
@@ -918,6 +931,85 @@ impl SourcePreferenceScope for SqliteStorage {
     }
 }
 
+#[derive(FromRow)]
+struct QueriedSourceReaction {
+    #[allow(dead_code)]
+    source: String,
+    #[allow(dead_code)]
+    weight: i32,
+    liked: bool,
+}
+
+#[async_trait]
+impl SourceReactionScope for SqliteStorage {
+    async fn fetch_source_reaction(&self, source: &str) -> Result<Option<bool>, Error> {
+        let mut tx = self.pool.begin().await?;
+
+        let reaction = sqlx::query_as::<_, QueriedSourceReaction>(
+            "SELECT source, weight, liked
+            FROM SourceReaction
+            WHERE source = ?;",
+        )
+        .bind(source)
+        .fetch_optional(&mut tx)
+        .await?;
+
+        tx.commit().await?;
+
+        Ok(reaction.map(|r| r.liked))
+    }
+
+    async fn store_source_reaction(&self, source: &str, liked: bool) -> Result<(), Error> {
+        let weight = if liked { 1 } else { -1 };
+        let last_updated = Utc::now().naive_utc();
+
+        let mut tx = self.pool.begin().await?;
+
+        sqlx::query(
+            "INSERT INTO SourceReaction (source, weight, lastUpdated, liked) VALUES (?, ?, ?, ?)",
+        )
+        .bind(source)
+        .bind(weight)
+        .bind(last_updated)
+        .bind(liked)
+        .execute(&mut tx)
+        .await?;
+
+        tx.commit().await.map_err(Into::into)
+    }
+
+    async fn update_source_weight(&self, source: &str) -> Result<(), Error> {
+        let last_updated = Utc::now().naive_utc();
+
+        let mut tx = self.pool.begin().await?;
+
+        sqlx::query(
+            "UPDATE SourceReaction
+            SET lastUpdated = ?,
+                weight = CASE WHEN liked = 1 THEN (weight + 1) END
```

Comment:
so we only increment when user likes and do nothing to the `weight` in case of a dislike?

Reply:
right, `weight` is incremented in the case of liked, untouched in the case of disliked.